libgnucash/engine/Account.cpp

```diff
@@ -3397,10 +3440,16 @@ GetBalanceAsOfDate (Account *acc, time64 date, gboolean ignclosing)
             /* Since lp is now pointing to a split which was past the reconcile
              * date, get the running balance of the previous split.
              */
-            if (ignclosing)
-                balance = xaccSplitGetNoclosingBalance( (Split *)lp->prev->data );
-            else
+            switch (baltype) {
+            case BalanceType::FULL:
                 balance = xaccSplitGetBalance( (Split *)lp->prev->data );
+                break;
+            case BalanceType::NO_CLOSING:
+                balance = xaccSplitGetNoclosingBalance( (Split *)lp->prev->data );
+                break;
+            case BalanceType::RECONCILED:
+                balance = xaccSplitGetReconciledBalance( (Split *)lp->prev->data );
```

Comment:
Close but not quite. The balances attached to the splits were computed with `xaccAccountRecomputeBalances` in `posted_date` order. If you're looking a couple of months back you'll have splits that were reconciled after the reconcile date of interest and so will be included in the reconciled balance computation.

Reply:
This whole clause also has an off-by-one error: If trans_time <= date then the balance of that split, not of the previous one, is the balance that you eventually want to return.